### DISCUSSION TITLE
Unable to access repository through GitHubClientProvider

### DISCUSSION THREAD

**Discussion by stianst** [quarkiverse/quarkus-github-app#566] [Q&A] [answered]
I'm trying to create a scheduled task that does some house keeping on old issues, but having some issues with invoking GH APIs.

I'm getting the GitHub instance from `@Inject GitHubClientProvider gitHubClientProvider; gitHubClientProvider.getApplicationClient()`, and that part is working fine. I can for example also use `GitHub#searchIssues`, but if I try `GitHub#getRepository` or anything else I'm getting a `bad credentials` message:

```
Caused by: org.kohsuke.github.HttpException: {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}
	at org.kohsuke.github.GitHubConnectorResponseErrorHandler$1.onError(GitHubConnectorResponseErrorHandler.java:62)
	at org.kohsuke.github.GitHubClient.detectKnownErrors(GitHubClient.java:473)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:434)
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:146)
	... 21 more
```

Pretty sure the application in GitHub has all the needed permissions, so not sure what's going on.

**Comment by gsmet** [comment, marked as answer]
The javadoc says it all :)
```
     * Gets the {@link GitHub GitHub client} for the application:
     * it can be used without any installation, but has very little access rights (almost as little as an anonymous client).
```

You cannot do anything useful with this client.

You will either need an installation client (but in this case, you need to push the installationId in the config as you won't be able to get it from a payload) or you need to inject a token and create a standard client with a token (using `new GitHubBuilder().withOAuthToken(the token).build()`).

Another option instead of pushing the installationId to the config might be to use something like `gitHubProvider.getApplicationClient().getApp().listInstallations()` or `getInstallationByOrganization()` or `getInstallationByRepository()` if you know which org/repo your installation is installed in.
Note that I haven't tested these methods with the app client so not sure if you have the permission.
Then you might be able to get an installation client from the installationId you get and the installation client has all the permissions you gave it.

If you need to create a config, it should be as easy as: https://github.com/quarkusio/quarkus-github-bot/blob/main/src/main/java/io/quarkus/bot/config/QuarkusGitHubBotConfig.java

Let me know if this solves your issue or if you have other questions.

**Reply by gsmet** [reply]
If you can get the installation thing working, I recommend it as it avoids having yet another token around.

**Reply by stianst** [reply]
Awesome, thanks and sorry for not reading the javadocs as that answers it all. Ended up doing what you suggested listing the installations to get the installation-id.

**Reply by gsmet** [reply]
No worries! The javadoc didn't give any details anyway. I will try to add something to the doc about it as it seems to be a common use case.
